### ISSUE TITLE
Can functions be called with arguments that are subtypes of the declared input arguments?

### ISSUE THREAD

**Issue by casella** [modelica/ModelicaSpecification#3103]
Consider the following MWE:
```modelica
package TestFunctionInputType
  record BaseR
    constant Integer n = 3;
    Real x[n];
  end BaseR;

  record DerivedR
    extends BaseR(n = 5);
  end DerivedR;

  function f
    input BaseR r;
    output Real y;
  algorithm
    y := sum(r);
  end f;

  model M
    DerivedR r(x={1, 2, 3, 4, 5});
    Real y;
  equation
    y = f(r);
  end M;
end TestFunctionInputType;
```
Is the function call in model M legal? The Specification, [Section 12.4.1](https://specification.modelica.org/master/functions.html#positional-or-named-input-arguments-of-functions) states:
> The type of each argument must agree with the type of the corresponding parameter

Strictly speaking, this is not the case, since `r` contains a record `x` of size 5, while the input argument of function `f` contains a record with size 3, so the types do not agree. For this reason, as reported by @perost, OMC rejects this MWE. Dymola instead accepts it.

My understanding is that, despite Modelica claiming to be an object-oriented languages, there is no such thing as polymorphism in Modelica functions, unlike OO languages like C++ or Java. The introduction of [Chapter 6](https://specification.modelica.org/master/interface-or-type-relationships.html), explicitly mentions the only cases where polymorphism is allowed: replaceable classes and inner/outer elements. This has nothing to do with function inputs, so I guess 12.4.1 should be interpreted strictly and model `M` should be rejected.

What is the interpretation of the language group?

**Comment by perost** [comment]
@eshmoylova: My original MWE was just:

**Comment by perost** [comment]
@eshmoylova: Newer versions of OpenModelica rejects it because `r` isn't an array, and `sum` requires the argument to be an array. But my original MWE was just: 
```Modelica
record BaseR
  constant Integer n = 3;
  Real x[n];
end BaseR;

record DerivedR
  extends BaseR(n = 5);
end DerivedR;

function f
  input BaseR r;
end f;

model M
  DerivedR r;
algorithm
  f(r);
end M;
```
It seems @casella added some bits to it in order to attempt to make it a balanced model (which I usually don't bother with for this kind of issue), but got it a bit wrong. But the body of the function doesn't matter, the question is about whether the argument is legal or not.

**Comment by casella** [comment]
@HansOlsson, @henrikt-ma, @perost, what do you think?

**Comment by perost** [comment]
> @HansOlsson, @henrikt-ma, @perost, what do you think?

The original issue in ClaRa was fixed, so there is no longer any use case for this. As such I can't see any reason why it should be allowed.

**Comment by casella** [comment]
I agree with @perost, but I'd like to get some feedback from other tool vendors if there is agreement on how to interpret the specification in this case.

**Comment by henrikt-ma** [comment]
My understanding of the Modelica type system is limited, but I wouldn't have expected this to be allowed.  On the other hand, it is the kind of matter where I wouldn't have been totally surprised if it turned out that tools are supposed to handle it by instantiating the function differently depending on the types of arguments passed to it.

**Comment by casella** [comment]
As I understand it, despite the claim the Modelica is an object-oriented language, it doesn't have polymorphism in function calls. The only place where polymorphism is allowed is redeclares. And maybe a little bit of it in operator records.

**Comment by eshmoylova** [comment]
I finally had a chance to look at it in detail. I am not sure DerivedR is not a type of BaseR. The dimensions of arrays must be the same, but how would you write that down? When generating code for function `f` with input `r` as:
```
  function f
    input BaseR r;
    ...
```
I would generate the equivalent of the following assertion:
```
assert(size(r.x,1) == r.n, "dimension mismatch");
```
And if `M.r` is passed to `f` it would still satisfy that assertion. So to me the question is: to which extent the dimensions must be evaluated when functions are generated?
